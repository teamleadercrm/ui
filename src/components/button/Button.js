import React, { useRef, forwardRef, useImperativeHandle } from 'react';
import PropTypes from 'prop-types';
import Box from '../box';
import LoadingSpinner from '../loadingSpinner';
import { UITextBody, UITextDisplay, UITextSmall } from '../typography';
import cx from 'classnames';
import theme from './theme.css';

const textComponentMap = {
  tiny: UITextSmall,
  small: UITextBody,
  medium: UITextBody,
  large: UITextDisplay,
};

const Button = forwardRef(
  (
    {
      color,
      level,
      onMouseUp,
      onMouseLeave,
      children,
      className,
      disabled,
      element,
      active,
      fullWidth,
      icon,
      iconPlacement,
      label,
      size,
      type,
      processing,
      ...others
    },
    ref,
  ) => {
    const buttonRef = useRef();
    useImperativeHandle(ref, () => buttonRef.current);

    const getSpinnerColor = () => {
      switch (level) {
        case 'secondary':
          return 'teal';
        case 'outline':
          return color === 'white' ? 'neutral' : color;
        case 'link':
          return 'aqua';
        default:
          return 'neutral';
      }
    };

    const getSpinnerTint = () => {
      switch (level) {
        case 'secondary':
          return 'darkest';
        case 'outline':
          return color === 'white' ? 'lightest' : 'darkest';
        case 'link':
          return 'dark';
        default:
          return 'lightest';
      }
    };

    const getSpinnerSize = () => {
      console.log(size);
      switch (size) {
        case 'tiny':
        case 'small':
          return 'small';
        default:
          return 'medium';
      }
    };

    const handleMouseUp = (event) => {
      blur();
      if (onMouseUp) {
        onMouseUp(event);
      }
    };

    const handleMouseLeave = (event) => {
      blur();
      if (onMouseLeave) {
        onMouseLeave(event);
      }
    };

    const blur = () => {
      const currentButtonRef = buttonRef.current;
      if (currentButtonRef) {
        currentButtonRef.blur();
      }
    };

    const classNames = cx(
      theme['reset-box-sizing'],
      theme['reset-font-smoothing'],
      theme['button-base'],
      theme['button'],
      theme[level],
      theme[size],
      {
        [theme['has-icon-only']]: (!children && !label) || (Array.isArray(children) && !children[0] && !label),
        [theme[color]]: level === 'outline',
        [theme['is-disabled']]: disabled,
        [theme['is-full-width']]: fullWidth,
        [theme['is-processing']]: processing,
        [theme['is-active']]: active,
      },
      className,
    );

    const props = {
      ...others,
      ref: buttonRef,
      className: classNames,
      disabled: element === 'button' ? disabled : null,
      element: element,
      onMouseUp: handleMouseUp,
      onMouseLeave: handleMouseLeave,
      type: element === 'button' ? type : null,
      'data-teamleader-ui': 'button',
    };

    const Text = textComponentMap[size];

    return (
      <Box {...props}>
        {icon && iconPlacement === 'left' && icon}
        {(label || children) && (
          <Text
            element="span"
            maxLines={1}
            marginLeft={icon && iconPlacement === 'left' && 2}
            marginRight={icon && iconPlacement === 'right' && 2}
          >
            {label}
            {children}
          </Text>
        )}
        {icon && iconPlacement === 'right' && icon}
        {processing && (
          <LoadingSpinner
            className={theme['spinner']}
            color={getSpinnerColor()}
            size={getSpinnerSize()}
            tint={getSpinnerTint()}
          />
        )}
      </Box>
    );
  },
);

Button.propTypes = {
  /** The content to display inside the button. */
  children: PropTypes.any,
  /** A class name for the button to give custom styles. */
  className: PropTypes.string,
  /** The color which the button should have when 'level' is set to 'outline' */
  color: PropTypes.oneOf(['teal', 'neutral', 'mint', 'violet', 'ruby', 'gold', 'aqua', 'white']),
  /** A custom element to be rendered */
  element: PropTypes.oneOfType([PropTypes.element, PropTypes.string]),
  /** Determines which kind of button to be rendered. */
  level: PropTypes.oneOf(['outline', 'primary', 'secondary', 'destructive', 'link', 'timer']),
  /** If true, component will be disabled. */
  disabled: PropTypes.bool,
  /** If true, component will be shown in an active state */
  active: PropTypes.bool,
  /** If true, component will take the full width available. */
  fullWidth: PropTypes.bool,
  /** The icon displayed inside the button. */
  icon: PropTypes.element,
  /** The position of the icon inside the button. */
  iconPlacement: PropTypes.oneOf(['left', 'right']),
  /** The textual label displayed inside the button. */
  label: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /** Callback function that is fired when mouse leaves the component. */
  onMouseLeave: PropTypes.func,
  /** Callback function that is fired when the mouse button is released. */
  onMouseUp: PropTypes.func,
  /** If true, component will show a loading spinner instead of label or children. */
  processing: PropTypes.bool,
  /** Size of the button. */
  size: PropTypes.oneOf(['tiny', 'small', 'medium', 'large']),
  /** Type of the button element. */
  type: PropTypes.string,
};

Button.defaultProps = {
  className: '',
  color: 'teal',
  element: 'button',
  fullWidth: false,
  level: 'secondary',
  iconPlacement: 'left',
  processing: false,
  size: 'medium',
  type: 'button',
};

Button.displayName = 'Button';

export default Button;
