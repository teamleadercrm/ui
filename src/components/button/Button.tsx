import cx from 'classnames';
import React, { forwardRef, ReactNode, useImperativeHandle, useRef } from 'react';
import { GenericComponent } from '../../@types/types';
import { COLORS, SIZES } from '../../constants';
import Box from '../box';
import { BoxProps } from '../box/Box';
import LoadingSpinner from '../loadingSpinner';
import { UITextBody, UITextDisplay, UITextSmall } from '../typography';
import theme from './theme.css';

export enum BUTTON_LEVELS {
  primary = 'primary',
  secondary = 'secondary',
  destructive = 'destructive',
  link = 'link',
  timer = 'timer',
  outline = 'outline',
}

const textComponentMap = {
  tiny: UITextSmall,
  small: UITextBody,
  medium: UITextBody,
  large: UITextDisplay,
};
export type AllowedButtonColor = typeof COLORS[number] | 'white';

export interface ButtonProps extends Omit<BoxProps, 'size'> {
  /** The content to display inside the button. */
  children?: ReactNode;
  /** A class name for the button to give custom styles. */
  className?: string;
  /** The color which the button should have when 'level' is set to 'outline' */
  color?: AllowedButtonColor;
  /** A custom element to be rendered */
  element?: any;
  /** If true, component will be disabled. */
  disabled?: boolean;
  /** If true, component will be shown in an active state */
  active?: boolean;
  /** If true, component will take the full width available. */
  fullWidth?: boolean;
  /** The icon displayed inside the button. */
  icon?: ReactNode;
  /** The position of the icon inside the button. */
  iconPlacement?: 'left' | 'right';
  /** The textual label displayed inside the button. */
  label?: string;
  /** Determines which kind of button to be rendered. */
  level?: `${BUTTON_LEVELS}`;
  /** If true, component will show a loading spinner instead of label or children. */
  processing?: boolean;
  /** Size of the button. */
  size?: Exclude<typeof SIZES[number], 'fullscreen' | 'smallest' | 'hero'>;
  /** Type of the button element. */
  type?: string;
  /** Button value in case it's being in a button group. */
  value?: string;
}

const Button: GenericComponent<ButtonProps> = forwardRef<HTMLElement, ButtonProps>(
  (
    {
      color = 'teal',
      children,
      className,
      disabled = false,
      element = 'button',
      active = false,
      fullWidth = false,
      icon,
      iconPlacement = 'left',
      label,
      level = BUTTON_LEVELS.secondary,
      size = 'medium',
      type = 'button',
      processing = false,
      ...others
    },
    ref,
  ) => {
    const buttonRef = useRef<HTMLElement>(null);
    useImperativeHandle<HTMLElement | null, HTMLElement | null>(ref, () => buttonRef.current);

    const getSpinnerColor = () => {
      switch (level) {
        case BUTTON_LEVELS.secondary:
          return 'teal';
        case BUTTON_LEVELS.outline:
          return color === 'white' ? 'neutral' : color;
        case BUTTON_LEVELS.link:
          return 'aqua';
        default:
          return 'neutral';
      }
    };

    const getSpinnerTint = () => {
      switch (level) {
        case BUTTON_LEVELS.secondary:
          return 'darkest';
        case BUTTON_LEVELS.outline:
          return color === 'white' ? 'lightest' : 'darkest';
        case BUTTON_LEVELS.link:
          return 'dark';
        default:
          return 'lightest';
      }
    };

    const getSpinnerSize = () => {
      switch (size) {
        case 'tiny':
        case 'small':
          return 'small';
        default:
          return 'medium';
      }
    };

    const classNames = cx(
      theme['reset-box-sizing'],
      theme['reset-font-smoothing'],
      theme['button-base'],
      theme['button'],
      theme[level],
      theme[size],
      {
        [theme['has-icon-only']]: (!children && !label) || (Array.isArray(children) && !children[0] && !label),
        [theme[color]]: level === BUTTON_LEVELS.outline,
        [theme['is-disabled']]: disabled,
        [theme['is-full-width']]: fullWidth,
        [theme['is-processing']]: processing,
        [theme['is-active']]: active,
      },
      className,
    );

    const Text = textComponentMap[size];

    return (
      <Box
        {...others}
        className={classNames}
        data-teamleader-ui="button"
        disabled={element === 'button' ? disabled : undefined}
        element={element}
        ref={buttonRef}
        type={element === 'button' ? type : undefined}
      >
        {icon && iconPlacement === 'left' && icon}
        {(label || children) && (
          <Text
            element="span"
            maxLines={1}
            marginLeft={icon && iconPlacement === 'left' ? 2 : undefined}
            marginRight={icon && iconPlacement === 'right' ? 2 : undefined}
          >
            {label}
            {children}
          </Text>
        )}
        {icon && iconPlacement === 'right' && icon}
        {processing && (
          <LoadingSpinner
            className={theme['spinner']}
            color={getSpinnerColor()}
            size={getSpinnerSize()}
            tint={getSpinnerTint()}
          />
        )}
      </Box>
    );
  },
);

Button.displayName = 'Button';

export default Button;
