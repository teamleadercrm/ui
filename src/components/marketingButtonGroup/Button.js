import React, { useRef, forwardRef, useImperativeHandle } from 'react';
import PropTypes from 'prop-types';
import Box from '../box';
import { UITextBody } from '../typography';
import cx from 'classnames';
import theme from './theme.css';

const Button = forwardRef(({ onMouseUp, onMouseLeave, children, className, active, label, ...others }, ref) => {
  const buttonRef = useRef();
  useImperativeHandle(ref, () => buttonRef.current);

  const handleMouseUp = (event) => {
    blur();
    if (onMouseUp) {
      onMouseUp(event);
    }
  };

  const handleMouseLeave = (event) => {
    blur();
    if (onMouseLeave) {
      onMouseLeave(event);
    }
  };

  const blur = () => {
    const currentButtonRef = buttonRef.current;
    if (currentButtonRef) {
      currentButtonRef.blur();
    }
  };

  const classNames = cx(
    theme['reset-box-sizing'],
    theme['reset-font-smoothing'],
    theme['button-base'],
    theme['button'],
    {
      [theme['is-active']]: active,
    },
    className,
  );

  const props = {
    ...others,
    ref: buttonRef,
    className: classNames,
    element: 'button',
    onMouseUp: handleMouseUp,
    onMouseLeave: handleMouseLeave,
    'data-teamleader-ui': 'button',
  };

  return (
    <Box {...props}>
      {(label || children) && (
        <UITextBody element="span" maxLines={1}>
          {label}
          {children}
        </UITextBody>
      )}
    </Box>
  );
});

Button.propTypes = {
  /** The content to display inside the button. */
  children: PropTypes.any,
  /** A class name for the button to give custom styles. */
  className: PropTypes.string,
  /** If true, component will be shown in an active state */
  active: PropTypes.bool,
  /** The textual label displayed inside the button. */
  label: PropTypes.oneOfType([PropTypes.number, PropTypes.string]),
  /** Callback function that is fired when mouse leaves the component. */
  onMouseLeave: PropTypes.func,
  /** Callback function that is fired when the mouse button is released. */
  onMouseUp: PropTypes.func,
};

Button.defaultProps = {
  className: '',
};

Button.displayName = 'MarketingButtonGroup.Button';

export default Button;
