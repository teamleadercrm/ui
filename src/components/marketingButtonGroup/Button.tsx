import React, { useRef, forwardRef, useImperativeHandle, ReactNode } from 'react';
import Box from '../box';
import { UITextBody } from '../typography';
import cx from 'classnames';
import theme from './theme.css';
import { BoxProps } from '../box/Box';
import { GenericComponent } from '../../@types/types';

export interface ButtonProps extends Omit<BoxProps, 'ref'> {
  /** The content to display inside the button. */
  children?: ReactNode;
  /** A class name for the button to give custom styles. */
  className?: string;
  /** If true, component will be shown in an active state */
  active?: boolean;
  /** The textual label displayed inside the button. */
  label?: number | string;
  /** Button value in case it's being in a button group. */
  value?: string;
}

const Button: GenericComponent<ButtonProps> = forwardRef<HTMLElement, ButtonProps>(
  ({ children, className = '', active, label, ...others }, ref) => {
    const buttonRef = useRef<HTMLElement>(null);
    useImperativeHandle<HTMLElement | null, HTMLElement | null>(ref, () => buttonRef.current);

    const classNames = cx(
      theme['reset-box-sizing'],
      theme['reset-font-smoothing'],
      theme['button-base'],
      theme['button'],
      {
        [theme['is-active']]: active,
      },
      className,
    );

    const props = {
      ...others,
      ref: buttonRef,
      className: classNames,
      element: 'button' as React.ElementType,
      'data-teamleader-ui': 'button',
    };

    return (
      <Box {...props}>
        {(label || children) && (
          <UITextBody element="span" maxLines={1}>
            {label}
            {children}
          </UITextBody>
        )}
      </Box>
    );
  },
);

Button.displayName = 'MarketingButtonGroup.Button';

export default Button;
