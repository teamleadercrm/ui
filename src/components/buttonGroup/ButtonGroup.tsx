import React, { PureComponent } from 'react';
import omit from 'lodash.omit';
import Box, { omitBoxProps, pickBoxProps } from '../box';
import Button from '../button';
import cx from 'classnames';
import isComponentOfType from '../utils/is-component-of-type';
import theme from './theme.css';
import { BoxProps } from '../box/Box';
import { BUTTON_LEVELS } from '../button/Button';
import isReactElement from '../utils/is-react-element';
import { SIZES } from '../../constants';

export interface ButtonGroupProps extends Omit<BoxProps, 'size' | 'onChange' | 'ref'> {
  /** The content to display inside the button group. */
  children?: React.ReactNode;
  /** A class name for the wrapper to give custom styles. */
  className?: string;
  /** If true, child components will be displayed segmented. */
  segmented?: boolean;
  /** The level of the buttons */
  size?: Exclude<typeof SIZES[number], 'fullscreen' | 'smallest' | 'hero'>;
  /** The level of the buttons */
  level?: keyof typeof BUTTON_LEVELS;
  /** The value of the currently active button. */
  value?: string;
  /** Callback function that is fired when the active button changes. */
  onChange?: (value: string, event: Event) => void;
}

class ButtonGroup extends PureComponent<ButtonGroupProps> {
  handleChange = (value: string, event: Event) => {
    if (this.props.onChange) {
      this.props.onChange(value, event);
    }
  };

  render() {
    const { children, className, segmented = false, value, ...others } = this.props;

    const classNames = cx(
      theme['group'],
      {
        [theme['segmented']]: segmented,
      },
      className,
    );

    return (
      <Box data-teamleader-ui="button-group" className={classNames} {...pickBoxProps(others)}>
        {React.Children.map(children, (child) => {
          if (!isReactElement(child) || (isReactElement(child) && !isComponentOfType(Button, child))) {
            return child;
          }

          let optionalChildProps = {};
          if (value) {
            optionalChildProps = {
              active: child.props.value === value,
              onClick: (event: Event) => this.handleChange(child.props.value, event),
            };
          }

          const childProps = omit(child.props, ['value']);
          const groupProps = omit(others, ['onChange']);

          return React.createElement(child.type, {
            ...childProps,
            ...optionalChildProps,
            ...omitBoxProps(groupProps),
          });
        })}
      </Box>
    );
  }
}

export default ButtonGroup;
