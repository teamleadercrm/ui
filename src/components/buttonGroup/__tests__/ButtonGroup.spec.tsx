import React from 'react';
import { render } from '@testing-library/react';
import ButtonGroup from '../ButtonGroup';
import Button from '../../Button';

describe('Component - ButtonGroup', () => {
  it('renders', async () => {
    const { asFragment } = render(
      <ButtonGroup value="button-1">
        <Button label="Button 1" value="button-1" />
        <Button label="Button 2" value="button-2" />
      </ButtonGroup>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
