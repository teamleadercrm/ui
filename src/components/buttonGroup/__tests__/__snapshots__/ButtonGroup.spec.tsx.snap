// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component - ButtonGroup renders 1`] = `
<DocumentFragment>
  <div
    class="box group"
    data-teamleader-ui="button-group"
  >
    <button
      class="box reset-box-sizing reset-font-smoothing button-base button secondary medium is-active"
      data-teamleader-ui="button"
      type="button"
    >
      <span
        class="box ui-text ui-text-body darkest overflow-singleline"
        data-teamleader-ui="ui-text"
      >
        Button 1
      </span>
    </button>
    <button
      class="box reset-box-sizing reset-font-smoothing button-base button secondary medium"
      data-teamleader-ui="button"
      type="button"
    >
      <span
        class="box ui-text ui-text-body darkest overflow-singleline"
        data-teamleader-ui="ui-text"
      >
        Button 2
      </span>
    </button>
  </div>
</DocumentFragment>
`;
